clients/vault/src/oracle/collector/proof_builder.rs

```diff
@@ -182,43 +182,90 @@ impl ScpMessageCollector {
 		tracing::debug!("Fetching SCP envelopes for slot {} from horizon archive", slot);
 		let envelopes_map_arc = self.envelopes_map_clone();
 
-		let stellar_history_base = self.stellar_history_base_url();
+		let archive_urls = self.stellar_history_archive_urls();
 		async move {
-			let scp_archive_storage = ScpArchiveStorage(stellar_history_base);
-			let scp_archive_result = scp_archive_storage.get_archive(slot).await;
-			if let Err(e) = scp_archive_result {
-				tracing::error!(
-					"Could not get SCPArchive for slot {slot:?} from Horizon Archive: {e:?}"
-				);
-				return
+			if archive_urls.is_empty() {
+				tracing::debug!("Can't get envelopes from horizon archive for slot {slot:?}: no archive URLs configured");
+				return Err("No archive URLs configured".to_string())
 			}
-			let scp_archive: XdrArchive<ScpHistoryEntry> = scp_archive_result.unwrap();
 
-			let value = scp_archive.get_vec().iter().find(|&scp_entry| {
-				if let ScpHistoryEntry::V0(scp_entry_v0) = scp_entry {
-					Slot::from(scp_entry_v0.ledger_messages.ledger_seq) == slot
-				} else {
-					false
+			// We try to get the SCPArchive from each archive URL until we succeed or run out of
+			// URLs
+			for archive_url in archive_urls {
+				let scp_archive_storage = ScpArchiveStorage(archive_url);
+				let scp_archive_result = scp_archive_storage.get_archive(slot).await;
+				if let Err(e) = scp_archive_result {
+					tracing::error!(
+						"Could not get SCPArchive for slot {slot:?} from Horizon Archive: {e:?}"
+					);
+					continue
 				}
-			});
-
-			if let Some(i) = value {
-				if let ScpHistoryEntry::V0(scp_entry_v0) = i {
-					let slot_scp_envelopes = scp_entry_v0.clone().ledger_messages.messages;
-					let vec_scp = slot_scp_envelopes.get_vec().clone();
-
-					let mut envelopes_map = envelopes_map_arc.write();
-
-					if envelopes_map.get(&slot).is_none() {
-						tracing::debug!("Adding archived SCP envelopes for slot {}", slot);
-						envelopes_map.insert(slot, vec_scp);
+				let scp_archive: XdrArchive<ScpHistoryEntry> =
+					scp_archive_result.expect("Should unwrap SCPArchive");
+
+				let value = scp_archive.get_vec().iter().find(|&scp_entry| {
+					if let ScpHistoryEntry::V0(scp_entry_v0) = scp_entry {
+						Slot::from(scp_entry_v0.ledger_messages.ledger_seq) == slot
+					} else {
+						false
+					}
+				});
+
+				if let Some(i) = value {
+					if let ScpHistoryEntry::V0(scp_entry_v0) = i {
+						let slot_scp_envelopes = scp_entry_v0.clone().ledger_messages.messages;
+						let vec_scp = slot_scp_envelopes.get_vec().clone();
+
+						// Filter out any envelopes that are not externalize or confirm statements
+						let relevant_envelopes = vec_scp
+							.into_iter()
+							.filter(|scp| match scp.statement.pledges {
+								ScpStatementPledges::ScpStExternalize(_) |
+								ScpStatementPledges::ScpStConfirm(_) => true,
+								_ => false,
+							})
+							.collect::<Vec<_>>();
+
+						let externalized_envelopes = relevant_envelopes
```

Comment:
Just a suggestion:
`externalized_envelopes` is not actually used; might be better if we use `filter(...).count()` instead? That way we're not collecting duplicate envelopes

Reply:
Good point 👍